packages/flutter_tools/lib/src/base/deferred_component.dart

```diff
@@ -0,0 +1,291 @@
+    if (allLoadingUnits == null) {
+      return;
+    }
+    for (final String lib in libraries) {
+      for (final LoadingUnit loadingUnit in allLoadingUnits) {
+        if (loadingUnit.libraries.contains(lib)) {
+          _loadingUnits.add(loadingUnit);
+        }
+      }
+    }
+  }
+
+  /// Provides a human readable string representation of the
+  /// configuration.
+  @override
+  String toString() {
+    String out = '\nDeferredComponent: $name\n  Libs:';
+    for (final String lib in libraries) {
+      out += '\n    $lib';
+    }
+    out += '\n  LoadingUnits:';
+    if (loadingUnits != null) {
+      for (final LoadingUnit loadingUnit in loadingUnits) {
+        out += '\n    ${loadingUnit.id}';
+      }
+    }
+    out += '\n  Assets:';
+    for (final Uri asset in assets) {
+      out += '\n    ${asset.path}';
+    }
+    return out;
+  }
+}
+
+/// Represents a single loading unit and holds information regarding it's id,
+/// shared library path, and dart libraries in it.
+class LoadingUnit {
+  /// Constructs a [LoadingUnit].
+  ///
+  /// Loading units must include an [id] and [libraries]. The [path] is only present when
+  /// parsing the loading unit from a laoding unit manifest produced by gen_snapshot.
+  LoadingUnit({
+    @required this.id,
+    @required this.libraries,
+    this.path,
+  });
+
+  /// The unique loading unit id that is used to identify the loading unit within dart.
+  final int id;
+
+  /// A list of dart libraries that the loading unit contains.
+  final List<String> libraries;
+
+  /// The output path of the shared library .so file created by gen_snapshot.
+  ///
+  /// This value may be null when the loading unit is parsed from a
+  /// `deferred_components_golden.yaml` file, which does not store the path.
+  final String path;
+
+  /// Returns a human readable string representation of this LoadingUnit, ignoring
+  /// the [path] field. The [path] is not included as it is not relevant when the
+  @override
+  String toString() {
+    String out = '\nLoadingUnit $id\n  Libraries:';
+    for (final String lib in libraries) {
+      out += '\n  - $lib';
+    }
+    return out;
+  }
+
+  /// Returns true if the other loading unit has the same [id] and the same set of [libraries],
+  /// ignoring order.
+  bool equalsIgnoringPath(LoadingUnit other) {
+    return other.id == id && other.libraries.toSet().containsAll(libraries);
+  }
+
+  /// Parses the loading unit manifests from the [Environment.outputDir] of the latest
+  /// gen_snapshot/assemble run.
+  ///
+  /// This will read all existing loading units for every provided abi. If no abis are
+  /// provided, loading units for all abis will be parsed.
+  static List<LoadingUnit> parseGeneratedLoadingUnits(Environment env, {List<String> abis}) {
+    final List<LoadingUnit> loadingUnits = <LoadingUnit>[];
+    final List<FileSystemEntity> files = env.outputDir.listSync(recursive: true);
+    while (files.isNotEmpty) {
```

Comment:
Is there a specific reason you're doing this while a while/removeLast? seems like a for-loop would be more staightforward.

Reply:
This was left over from a previous implementation where this was necessary. I have changed it to a for loop now.